### PR TITLE
TIP-229: Rework product creation popin to remove old PEF forms

### PR DESCRIPTION
**Pull request by phaseinducer**
**Description (for Contributor and Core Developer)**

Product creation popin was the last place in the PIM where we needed old Product form types.
This PR proposes a rework of the popin in js so we can finally burn the forms down and move forward.

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | n/a
| Added Behats                      | n/a
| Changelog updated                 |
| Review and 2 GTM                  |
| Micro Demo to the PO (Story only) |
| Migration script                  | n/a
| Tech Doc                          | n/a

### CODE REVIEW

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/ProductController.php [hunk 1]
```diff
@@ -160,6 +160,36 @@ public function getAction($id)
         ));
     }
 
+    public function createAction(Request $request)
```
Comment (skeleton) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/ProductController.php]: phpdoc

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/ProductController.php [hunk 2]
```diff
@@ -160,6 +160,36 @@ public function getAction($id)
...
+            $request->request->get('identifier'),
+            $request->request->get('family', null)
+        );
+
+        $violations = $this->validator->validate($product);
+        if (0 === $violations->count()) {
+            $this->productSaver->save($product);
+
+            $normalizationContext = $this->userContext->toArray() + [
+                    'filter_types'               => ['pim.internal_api.product_value.view'],
```
Comment (ClementGautier) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/ProductController.php]: indent issue ?

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/ProductController.php [hunk 3]
```diff
@@ -160,6 +160,36 @@ public function getAction($id)
...
+                    'disable_grouping_separator' => true
+                ];
+
+            return new JsonResponse($this->normalizer->normalize(
+                $product,
+                'internal_api',
+                $normalizationContext
+            ));
+        } else {
```
Comment (ClementGautier) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/ProductController.php]: the else expression here is useless as you return in the previous if

**Comment by skeleton** [comment]
GTM once fixed :+1:

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/create/create.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        'backbone',
+        'routing',
+        'pim/form-builder',
+        'pim/user-context',
+        'oro/translator',
+        'oro/loading-mask',
+        'oro/navigation'
+    ],
+    function ($, _, Backbone, Routing, FormBuilder, UserContext, __, LoadingMask, Navigation) {
+        var CreateProductView =  Backbone.View.extend({
```
Comment (ClementGautier) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/create/create.js]: Why don't extends Backbone.BootstrapModal directly instead ?

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/create/form.js [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+                                    text: i18n.getLabel(value.labels, UserContext.get('catalogLocale'), value.code)
+                                });
+                            });
+
+                            return data;
+                        }
+                    }
+                };
+
+                initSelect2.init(this.$('[data-code="family"] input'), options);
```
Comment (ClementGautier) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/create/form.js]: isn't this perform a request inside the whole page DOM ? could we make a better selector using this.$el as scope ?

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/attribute-option/form.js [hunk 6]
```diff
@@ -31,8 +26,8 @@ define(
                     };
                 });
 
-                this.model.set('code', this.$('input[name="code"]').val());
-                this.model.set('optionValues', optionValues);
+                this.getFormModel().set('code', this.$('input[name="code"]').val());
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/attribute-option/form.js]: 👍

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/create/create.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+                    content: '',
+                    cancelText: __('pim_enrich.entity.product.create_popin.labels.cancel'),
+                    okText: __('pim_enrich.entity.product.create_popin.labels.ok'),
+                    okCloses: false
+                });
+
+                modal.open();
+
+                var modalBody = modal.$('.modal-body');
+                modalBody.css('min-height', 150);
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/create/create.js]: no way to do that in css ?

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/create/create.js [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+                                    ));
+                                });
+                        });
+                    }.bind(this));
+
+                return deferred.promise();
+            }
+        });
+
+        return function () {
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/create/create.js]: I don't really understand why you do that. Can you explain? IRL is maybe better

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/ProductController.php (hunk 3 above)
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/ProductController.php]: Right!

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/create/form.js (hunk 5 above)
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/create/form.js]: This is already the behavior when you use `this.$`

File: src/Pim/Bundle/UIBundle/Resources/public/css/select2.less [hunk 9]
```diff
@@ -276,3 +276,9 @@ form.product-edit {
     width: 160px;
   }
 }
+
+.modal {
```
Comment (juliensnz) [inline src/Pim/Bundle/UIBundle/Resources/public/css/select2.less]: maybe too generic?

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/create/create.js (hunk 7 above)
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/create/create.js]: I wish, but the modal class is hardcoded here : https://github.com/akeneo/pim-community-dev/blob/master/src/Pim/Bundle/UIBundle/Resources/public/lib/backbone.bootstrap-modal.js#L50

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/create/create.js (hunk 4 above)
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/create/create.js]: I don't understand what would be the point :/

File: src/Pim/Bundle/UIBundle/Resources/public/css/select2.less (hunk 9 above)
Comment (phaseinducer) [inline src/Pim/Bundle/UIBundle/Resources/public/css/select2.less]: See my comment about class above.

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/ProductController.php (hunk 2 above)
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/ProductController.php]: Why that?

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/create/form.js (hunk 5 above)
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/create/form.js]: this.$('something') is exactly like doing this.$el.find('something')

**Comment by juliensnz** [comment]
GTM

**Comment by ClementGautier** [comment]
GTM

**Pull request merged**